[23.1] Cgroups initialization cycle fix for unpatched JDK 21+35

jerboaa opened the pull request:
Break the initialization cycle in NIO/cgroups code
    
First, we use a separate accessor for page-alignedness as it doesn't
need the more sophisticated initialization of the directMemory field.
    
Next, ensure PhysicalMemory initialization is serialized and when it is,
set directMemory to a static value so that the container code can finish
initialization without introducing a cyle. The final directMemory value
based on the heap size is then published to JDK code by setting the VM
init level to 1. Therefore, application code would use the non-static
value as the upper bound.
    
Closes: #556

jerboaa committed:
Break the initialization cycle in NIO/cgroups code

First, we use a separate accessor for page-alignedness as it doesn't
need the more sophisticated initialization of the directMemory field.

Next, ensure PhysicalMemory initialization is serialized and when it is,
set directMemory to a static value so that the container code can finish
initialization without introducing a cyle. The final directMemory value
based on the heap size is then published to JDK code by setting the VM
init level to 1. Therefore, application code would use the non-static
value as the upper bound.

Closes: #556

jerboaa committed:
Fixes for -XX:MaxRAM and -XX:MaxDirectMemorySize at runtime

zakkak commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/heap/PhysicalMemory.java:
```diff
@@ -78,24 +90,42 @@ public static UnsignedWord size() {
-                                    : ImageSingletons.lookup(PhysicalMemorySupport.class).size();
+        synchronized (INITIALIZING) {
+            if (!isInitialized()) {
+                VMError.guarantee(INITIALIZING.get() <= 1, "Must not initialize twice");
```
This seems redundant. Since the code is synchronized on INITIALIZING and `isInitialized()` is `false`, it is guaranteed that `INITIALIZING.get() <= 1` will be true, but even if it wasn't this check would be pretty "racy".

jerboaa replied:
Removed now.

zakkak commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/heap/PhysicalMemory.java:
```diff
@@ -78,24 +90,42 @@ public static UnsignedWord size() {
+                    // available to use.
+                    CACHED_SIZE_AVAIL_LATCH.countDown();
+                } finally {
+                    INITIALIZING.incrementAndGet();
```
My understanding is that this will run even if line 118 throws an exception and `cachedSize` won't get set. 

Shouldn't we just fail initialization and return the state to uninitialized so that other parts can handle the issue? E.g. https://github.com/oracle/graal/blob/544b6f80d090bc320b69318efb325075de87e2a9/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_jdk_internal_misc_VM.java#L140

Alternatively shouldn't we set a fallback value for `cachedSize` and count down the latch in this case?

jerboaa replied:
This should be fixed in the updated version.

zakkak reviewed (changes requested):
Please see the comments and https://github.com/jerboaa/graal/pull/3 for a suggestion on how to achieve the same without a `synchronized` block.

zakkak commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_jdk_internal_misc_VM.java:
```diff
@@ -67,36 +70,50 @@ public static ClassLoader latestUserDefinedLoader0() {
+    @Alias @InjectAccessors(PageAlignDirectMemoryAccessors.class) //
     private static boolean pageAlignDirectMemory;
+
+    @Alias //
+    @SuppressWarnings("unused")
+    public static void initLevel(int newVal) {
+        throw VMError.shouldNotReachHere("This is an alias to the method in the target class, so this code is unreachable");
+    }
+
+    @Alias //
+    @SuppressWarnings("unused")
+    public static int initLevel() {
+        return -1; // Aliased code. Should not reach here
+    }
```
Nitpick

```suggestion
    @Alias
    public static native void initLevel(int newVal);

    @Alias
    public static native int initLevel();
```

jerboaa replied:
Fixed.

jerboaa committed:
Use ReentrantLock for PhysicalMemory synchronization

Don't use volatile in DirectMemoryAccessors.isInitialized and don't use
synchronization as the outcome of those races should amount to the same
values.

zakkak commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/heap/PhysicalMemory.java:
```diff
@@ -78,24 +86,42 @@ public static UnsignedWord size() {
+                    throw VMError.shouldNotReachHere("Interrupt on countdown latch!");
+                }
+            }
+            if (INITIALIZATION_STATE.compareAndSet(NOT_INITIALIZED, INITIALIZING)) {
```
After giving it a bit more thought, there is still a race we don't handle. If thread 1 sees that the state is `NOT_INITIALIZED` it will skip the previous `if` statement and jump right here. In the meantime if another thread 2 has started the initialization process the `compareAndSet` will fail here and Thread 1 will continue without an getting the proper `cachedSize`.

zakkak replied:
Proposed fix in https://github.com/jerboaa/graal/pull/4

jerboaa replied:
Correct and I think I've reproduced this issue in one of my tests:

```
$ cat MaxMemTest.java 
public class MaxMemTest {
	public static void main(String[] args) {
		long maxMem = Runtime.getRuntime().maxMemory();
		System.out.println("Heap max       = " + maxMem);
		System.out.println("PhysicalMemory = " + (long)(maxMem / 0.8));
	}
}
```

When I run that with:

```
$ ./maxmemtest -XX:MaxDirectMemorySize=10
Heap max       = 281474976186368
PhysicalMemory = 351843720232960
```

Should give:
```
$ ./maxmemtest -XX:MaxDirectMemorySize=10
java.lang.OutOfMemoryError: Cannot reserve 8192 bytes of direct buffer memory (allocated: 0, limit: 10)
    at java.nio.Bits.reserveMemory(Bits.java:178)
    at java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:127)
    at java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:360)
    at sun.nio.ch.Util.getTemporaryDirectBuffer(Util.java:242)
    at sun.nio.ch.IOUtil.read(IOUtil.java:303)
    at sun.nio.ch.IOUtil.read(IOUtil.java:283)
    at sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:234)
    at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:74)
    at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103)
    at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:329)
    at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:372)
    at sun.nio.cs.StreamDecoder.lockedRead(StreamDecoder.java:215)
    at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:169)
    at java.io.InputStreamReader.read(InputStreamReader.java:188)
    at java.io.BufferedReader.fill(BufferedReader.java:160)
    at java.io.BufferedReader.implReadLine(BufferedReader.java:370)
    at java.io.BufferedReader.readLine(BufferedReader.java:347)
    at java.io.BufferedReader.readLine(BufferedReader.java:436)
    at java.nio.file.Files.readAllLines(Files.java:3395)
    at java.nio.file.Files.readAllLines(Files.java:3433)
    at jdk.internal.platform.CgroupUtil.lambda$readAllLinesPrivileged$2(CgroupUtil.java:83)
    at java.security.AccessController.executePrivileged(AccessController.java:114)
    at java.security.AccessController.doPrivileged(AccessController.java:571)
    at jdk.internal.platform.CgroupUtil.readAllLinesPrivileged(CgroupUtil.java:84)
    at jdk.internal.platform.CgroupSubsystemFactory.determineType(CgroupSubsystemFactory.java:143)
    at jdk.internal.platform.CgroupSubsystemFactory.create(CgroupSubsystemFactory.java:85)
    at jdk.internal.platform.CgroupMetrics.getInstance(CgroupMetrics.java:193)
    at jdk.internal.platform.SystemMetrics.instance(SystemMetrics.java:29)
    at jdk.internal.platform.Metrics.systemMetrics(Metrics.java:58)
    at jdk.internal.platform.Container.metrics(Container.java:43)
    at com.oracle.svm.core.Containers.memoryLimitInBytes(Containers.java:121)
    at com.oracle.svm.core.heap.PhysicalMemory.size(PhysicalMemory.java:118)
    at com.oracle.svm.core.graal.snippets.CEntryPointSnippets.initializeIsolate(CEntryPointSnippets.java:375)
    at com.oracle.svm.core.JavaMainWrapper$EnterCreateIsolateWithCArgumentsPrologue.enter(JavaMainWrapper.java:447)
```

jerboaa replied:
I've switched to synchronization using `ReentrantLock`, which I think makes this code correct. Please let me know what you think.

jerboaa commented:
@zakkak Please take another look. Thanks!

zakkak reviewed (approved):
LGTM

jerboaa commented:
Thanks for the review! CI looks good, so I'll merge.

jerboaa commented:
I'll also propose this patch upstream at some point (probably next week).

zakkak commented:
> I'll also propose this patch upstream at some point (probably next week).

Yes, we need it for 24.0 as well, so getting it upstream would be the best.
